code/modules/chemistry/Chemistry-Machinery.dm

```diff
@@ -291,6 +291,183 @@ TYPEINFO(/obj/machinery/chem_heater)
+/obj/machinery/chem_shaker
+	name = "\improper Orbital Shaker"
+	desc = "A machine which continuously agitates beakers and flasks when activated."
+	icon = 'icons/obj/shaker.dmi'
+#ifdef IN_MAP_EDITOR
+	icon_state = "orbital_shaker-map"
+#else
+	icon_state = "orbital_shaker"
+#endif
+	density = 1
+	anchored = ANCHORED
+	flags = NOSPLASH
+	deconstruct_flags = DECON_SCREWDRIVER | DECON_WRENCH
+	pixel_y = 4
+
+	var/list/obj/item/reagent_containers/glass/held_containers = list()
+	var/obj/dummy/platform_holder
+	var/list/first_container_offsets = list("X" = 0, "Y" = 8)
+	var/list/container_offsets = list ("X" = 10, "Y" = -4)
+	var/active = FALSE
+	/// The arrangement of the containers on the platform in the X direction.
+	var/container_row_length = 1
+	/// Also acts as the number of containers in the Y direction when divided by `src.container_row_length`.
+	var/max_containers = 1
+	/// The time it takes for the platform to complete one orbit.
+	var/orbital_period = 0.6 SECONDS
+	/// Radius of the platform's orbit in pixels.
+	var/radius = 2
+
+	New()
+		..()
+		src.platform_holder = new()
+		src.platform_holder.icon = src.icon
+		src.platform_holder.icon_state = "[src.icon_state]-platform"
+		src.platform_holder.vis_flags |= VIS_INHERIT_ID | VIS_INHERIT_LAYER | VIS_INHERIT_PLANE
+		src.platform_holder.appearance_flags |= KEEP_TOGETHER
+		src.vis_contents.Add(src.platform_holder)
+
+	disposing()
+		for (var/obj/item/reagent_containers/glass/B in src.held_containers)
+			MOVE_OUT_TO_TURF_SAFE(B, src)
+		..()
+
+	attack_hand(mob/user)
+		if (!can_act(user)) return
+		boutput(user, "You [src.active ? "de" : ""]activate [src].")
+		switch (src.active)
+			if (TRUE)
+				src.set_inactive()
+			if (FALSE)
+				src.set_active()
+
+	attackby(obj/item/reagent_containers/glass/B, var/mob/user)
+		if(istype(B, /obj/item/reagent_containers/glass))
+			src.try_insert(B, user)
+
+	ex_act(severity)
+		switch(severity)
+			if(1)
+				qdel(src)
+				return
+			if(2)
+				if (prob(50))
+					qdel(src)
+					return
+
+	blob_act(var/power)
+		if(prob(25 * power/20))
+			qdel(src)
+
+	meteorhit()
+		qdel(src)
+		return
+
+	attack_ai(mob/user as mob)
+		return src.Attackhand(user)
+
+	process(mult)
+		..()
+		if (src.status & (NOPOWER|BROKEN)) return src.set_inactive()
+		for (var/obj/item/reagent_containers/glass/container in src.held_containers)
+			container.reagents?.physical_shock(5)
+
+	proc/arrange_containers()
+		if (!src.count_held_containers()) return
+		for (var/i in 1 to length(src.held_containers))
+			if (!src.held_containers[i]) continue
+			var/current_y = ceil(i / src.container_row_length)
+			var/current_x = i - (src.container_row_length * (current_y - 1))
+			src.held_containers[i].pixel_x = src.first_container_offsets["X"] + ((current_x - 1) * src.container_offsets["X"])
+			src.held_containers[i].pixel_y = src.first_container_offsets["Y"] + ((current_y - 1) *src.container_offsets["Y"])
+
+	proc/count_held_containers()
```

Comment:
> Do I dare ask why this isn't just `length()`? Should probably be managing null refs in some other way.

Hooh boy, I was afraid of this conversation.

The reason why beaker indexes get nulled instead of removing them directly from the list was due to the fact that the pixel offset of each container on the platform is calculated with that beaker's index in the `held_containers` list. If newly added beakers were removed from the list and changing the indexes of each one still remaining, it'd start reshuffling containers around when they shouldn't reasonably be. Unless there are better ways of tracking beaker positions properly, this was about the only method that I reasoned would work for the effect I wanted.

Reply:
That makes sense actually, it's just rare that treating a byond list like a regular array is the right thing to do.